### PR TITLE
Add mio based server

### PR DESCRIPTION
**Pull request by tisonkun**
Signed-off-by: tison <wander4096@gmail.com>

### CODE REVIEW

**Comment by huachaohuang** [comment]
You can try to set the `NO_DELAY` flag for connections, it would help in some cases.

**Comment by tisonkun** [comment]
@huachaohuang updated.

### tokio baseline

```
$ redis-benchmark -t set,get -n 1000 -q -c 1 -P 3 -p 21716
ERROR: ERR unknown command 'config'
ERROR: failed to fetch CONFIG from 127.0.0.1:21716
WARN: could not fetch server CONFIG
SET: 20448.98 requests per second, p50=0.079 msec       
GET: 25692.31 requests per second, p50=0.063 msec
```

### mio round 1

```
$ redis-benchmark -t set,get -n 1000 -q -c 1 -P 3 -p 21716
ERROR: ERR unknown command 'config'
ERROR: failed to fetch CONFIG from 127.0.0.1:21716
WARN: could not fetch server CONFIG
SET: 10020.00 requests per second, p50=0.239 msec       
GET: 9823.53 requests per second, p50=0.247 msec
```

### mio round 2 (reduce log)

```
$ redis-benchmark -t set,get -n 1000 -q -c 1 -P 3 -p 21716
ERROR: ERR unknown command 'config'
ERROR: failed to fetch CONFIG from 127.0.0.1:21716
WARN: could not fetch server CONFIG
SET: 31312.50 requests per second, p50=0.087 msec       
GET: 34551.73 requests per second, p50=0.079 msec
```

### mio round 3 (TCP_NODELAY)

```
$ redis-benchmark -t set,get -n 1000 -q -c 1 -P 3 -p 21716
ERROR: ERR unknown command 'config'
ERROR: failed to fetch CONFIG from 127.0.0.1:21716
WARN: could not fetch server CONFIG
SET: 31312.50 requests per second, p50=0.087 msec       
GET: 38538.46 requests per second, p50=0.079 msec
```

It seems the improvements is not significant. When I increase `n` to 1,000,000, the result doesn't change.

File: src/server/src/uio/connection.rs [hunk 1]
```diff
@@ -27,9 +27,9 @@ pub struct Connection {
     fd: RawFd,
     io: IoDriver,
     db: Db,
-    rbuf: ReadBuf,
-    wbuf: WriteBuf,
-    num_inflights: usize,
+    read_buf: ReadBuf,
+    write_buf: WriteBuf,
+    inflight_cnt: usize,
```
Comment (huachaohuang) [inline src/server/src/uio/connection.rs:32]: Hmm, I prefer `num_xxx` than `xxx_cnt`.

File: src/bin/src/server.rs [hunk 2]
```diff
@@ -35,19 +35,41 @@ enum SubCommand {
     Start(StartCommand),
 }
 
+#[derive(clap::ArgEnum, Clone)]
+enum DriverMode {
+    Tokio,
```
Comment (huachaohuang) [inline src/bin/src/server.rs:40]: Seems the `mio` mode performs much better, should we remove the `tokio` mode?

**Review by huachaohuang — APPROVED** [review]
Rest LGTM. BTW, I think there are some problems with the buffer management, which causes connection resets or panic in some cases. Maybe @zojw is working on those parts.

File: src/server/src/uio/connection.rs (hunk 1 above)
Comment (tisonkun) [reply src/server/src/uio/connection.rs:32]: OK. Let me revert this change.

File: src/bin/src/server.rs (hunk 2 above)
Comment (tisonkun) [reply src/bin/src/server.rs:40]: Let's do it in a separated PR then.

**Comment by tisonkun** [comment]
> which causes connection resets or panic in some cases

@huachaohuang yes. Also I notice some cases that can return to the client an `ERR` instead of server crashing. We can handle errors properly in several small patches.

**Comment by huachaohuang** [comment]
A future enhancement: https://github.com/engula/engula/issues/658

File: src/server/src/uio/connection.rs [hunk 3]
```diff
@@ -27,9 +27,9 @@ pub struct Connection {
     fd: RawFd,
     io: IoDriver,
     db: Db,
-    rbuf: ReadBuf,
-    wbuf: WriteBuf,
-    num_inflights: usize,
+    read_buf: ReadBuf,
+    write_buf: WriteBuf,
+    num_inflghts: usize,
```
Comment (huachaohuang) [inline src/server/src/uio/connection.rs:32]:
```suggestion
    num_inflights: usize,
```

Comment (tisonkun) [reply src/server/src/uio/connection.rs]: Fixed.

**Pull request merged**